### DISCUSSION TITLE
Making the Tools fully configurable under Ubuntu

### DISCUSSION THREAD

**Discussion by Toby-Valtech** [actions/virtual-environments#3535] [Ideas]
Hi,

we are using the resources in this Repository to build our ubuntu agents for azure devops. This is done via Pipeline, however we deemed the ammount of installed software uneccessarily high and decided to remove some of it before running packer. This is done as json manipulation via script and mostly removes data from ubuntu-2004.json and toolset.json.

However in the file Install-Toolset.ps1:31 three more tools are declared ( `$toolsToInstall = @("Python", "Node", "Go")`  ), all of which need to be installed. We would like for example not to install the google-cloud-sdk, however we cant remove it without the pipeline falining.

Is there a specific reason why this piece of configuration can not be put into either a new or already existing json config file? If there is not would you accept a PR in which I implement this change?

Kudos Toby

**Comment by miketimofeev** [comment]
Hi @Toby-Valtech!
What exact configuration do you mean? Could you please provide some examples?
We're thinking of providing smaller images, but no ETA for now.

**Reply by Toby-Valtech** [reply]
Okay, currently as I understand it, most of the configuration regarding what is installed in the image is done in two places: Ubuntu-2004.json and toolsets/toolset-2004.json (Obviously this is in regard to the Ubuntu 20.04 Image). 

Because we want to remove software we dont need, to speed up the build process and shrink the image, I wrote a script to remove lines from these two files.

For example I wrote a whitelist of programms I want to keep.
Then I filter the Ubuntu-2004.json script to remove all but those entries from the block where all of the programms are specified [Line 195-251 in ubuntu-2004.json]. I attached our final toolset and ubuntu2004.json, so you can see the changes in action. [linux.zip](https://github.com/actions/virtual-environments/files/6607243/linux.zip)

However there is one piece of configuration that is not part of these two files. In the file .images/linux/scripts/installers/Configure-Toolset.ps1 in line 31. The tools to install are specified againm this time as `$toolsToInstall = @("Python", "Node", "Go")`. We dont really want to install either Python or Go, however we are forced to do so, because of this line. My question is thus if its possible to put this configuration (`"Python", "Node","Go"`) into another json file somwhere, so it can be edited with json manipulation.

I would be willing to make this change myself and open up a PR for it, if you guys accept external pull requests at all

Cheers Toby

**Reply by miketimofeev** [reply]
@Toby-Valtech now I've got your idea, thanks for the explanation. Yes, we accept contributions, feel free to open a PR

**Comment by miketimofeev** [comment]
@Toby-Valtech thanks for provided PR, I'm going to lock this conversation as the initial issue is resolved. Please feel free to contact us if you have any concerns.
